BZ-42472: feat: Implement UPI payment redirect handling in BlazeWebView

punyamsingh opened the pull request:
## Jira Ticket
refer [this](https://juspay.atlassian.net/browse/BZ-42472)

## Why?

- Enable seamless UPI payment experience for Android users by automatically redirecting UPI payment URLs to their respective payment apps  
- Prevent UPI payment links from loading within the WebView, which can cause payment failures or poor user experience  
- Support major Indian payment applications to ensure broad compatibility across the payment ecosystem  
- Improve payment conversion rates by providing native app experiences for UPI transactions  
- Achieve feature parity with iOS BlazeSDK UPI redirect implementation

## Solution

- Created BlazeConstants.kt with centralized UPI_SCHEMES constant
- Added isUPIIntentUri() function to BlazeUtils.kt with strict UPI parameter validation
- Refactored BlazeWebView to use inline blazeWebViewClient object instead of default WebViewClient
- Enhanced WebView URL interception with proper UPI scheme detection
- Modernized code with Kotlin extensions (toUri(), sharedPreferences.edit{})
- Added missing imports for WebResourceRequest and WebViewClient
- Improved type safety by replacing Uri.parse() calls with toUri() extension

## How to Test?

1. Load the BlazeSDK integrated app on a test device (physical device recommended)
2. Navigate to a payment method that offers UPI payment options (e.g., Simpl Pay-in-3)
3. Complete the login process for the payment service
4. In the payment flow, select and open a UPI payment method (e.g., GPay)
5. Verify that the external UPI app opens successfully (indicating UPI redirect is working)
6. Test with other available UPI payment options in the flow
7. Ensure the payment process completes successfully in the external app
8. Verify that non-UPI navigation within the WebView continues to work normally

## Dev Proof
https://drive.google.com/file/d/17-zwclCaW2AKhvcbDyMlk9bimZxrrKDK/view?usp=sharing

## Priority
High

murdore commented on blaze/src/main/java/in/breeze/blaze/BlazeWebView.kt:
```diff
@@ -248,4 +249,23 @@ internal class BlazeWebView @SuppressLint(
+      val url = request?.url
+      val scheme = url?.scheme?.lowercase()
+      
+      val upiSchemes = setOf(
```
have we added package whitelist as part of build process? else it will not woek

https://stackoverflow.com/questions/74693187/android-how-to-open-upi-payment-apps-using-intent-chooser-not-working-on-andro

punyamsingh replied:
i have updated and added intent queries in Manifest file for this now.

punyamsingh commented on blaze/src/main/java/in/breeze/blaze/BlazeWebView.kt:
```diff
@@ -248,4 +249,23 @@ internal class BlazeWebView @SuppressLint(
-}
\ No newline at end of file
+  private inner class BlazeWebViewClient : WebViewClient() {
+    override fun shouldOverrideUrlLoading(view: WebView?, request: WebResourceRequest?): Boolean {
```
added the changes

punyamsingh commented on blaze/src/main/java/in/breeze/blaze/BlazeWebView.kt:
```diff
@@ -248,4 +249,23 @@ internal class BlazeWebView @SuppressLint(
+        "amazonpay", "mobikwik", "freecharge", "credpay"
+      )
+      
+      if (scheme != null && upiSchemes.contains(scheme)) {
```
made the change

sinha-sahil commented on blaze/src/main/AndroidManifest.xml:
```diff
@@ -8,5 +8,45 @@
           android:pathPattern=".*"
           android:scheme="upi" />
     </intent>
+    <intent>
```
Not part of SDK. revert

punyamsingh commented on blaze/src/main/java/in/breeze/blaze/BlazeWebView.kt:
```diff
@@ -248,4 +256,42 @@ internal class BlazeWebView @SuppressLint(
+  }
+
+  private inner class BlazeWebViewClient : WebViewClient() {
+    override fun shouldOverrideUrlLoading(view: WebView?, request: WebResourceRequest?): Boolean {
```
made the changes

punyamsingh commented on blaze/src/main/java/in/breeze/blaze/BlazeWebView.kt:
```diff
@@ -248,4 +256,42 @@ internal class BlazeWebView @SuppressLint(
\ No newline at end of file
+  private fun isIntentUPILink(uri: Uri): Boolean {
+      val scheme = uri.scheme?.lowercase() ?: return false
+      if (!uri.toString().lowercase().contains("pay")) return false
```
resolved

punyamsingh committed:
BZ-42472: feat: Implement UPI payment redirect handling in BlazeWebView

- Created BlazeConstants.kt with centralized UPI_SCHEMES constant
- Added isUPIIntentUri() function to BlazeUtils.kt with strict UPI parameter validation
- Refactored BlazeWebView to use inline blazeWebViewClient object instead of default WebViewClient
- Enhanced WebView URL interception with proper UPI scheme detection
- Modernized code with Kotlin extensions (toUri(), sharedPreferences.edit{})
- Added missing imports for WebResourceRequest and WebViewClient
- Improved type safety by replacing Uri.parse() calls with toUri() extension
